[JIT] optimize mutableType calls

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#35474 [JIT] optimize mutableType calls**

I had previously tried to optimize getMutableTypePtr calls by not recursing through container types, but it turns out there are a few uses of container types which refine their contained elements.
This attempt was in #35301

Now I am optimizing calls by caching TypePtr -> Mutable TypePtr conversions. Now that we are doing caching none of the functions marked as const are really const anymore. Previously many of the const functions actually mutated internal state, such as rebuildWriteCache.

one kind of annoying thing is that there is a general api for querying mutability isMutableType that doesn't use the cache, and one internal that does, isMutableTypeInternal. It would be nice if I could call isMutableType within alias analysis and it would dispatch to the internal function, but I'm not sure how to do that.

getMutableTypePtr showed up as 12% of the first run of FairSeq, so this is a function worth optimizing.

Differential Revision: [D20873493](https://our.internmc.facebook.com/intern/diff/D20873493)

Commit:
optimize mutableType calls

suo commented on torch/csrc/jit/ir/alias_analysis.cpp:
```diff
@@ -7,53 +7,114 @@
+      : mutable_type_cache_(mutable_type_cache) {}
+
+  c10::optional<TypePtr> getMutableType(const TypePtr& type) {
+    if (mutable_type_cache_) {
```
maybe take mutable_type_cache_ by ref, so you don't have to worry about nullity

suo commented on torch/csrc/jit/ir/alias_analysis.cpp:
```diff
@@ -7,53 +7,114 @@
+    std::unordered_map<TypePtr, TypePtr>* mutable_type_cache) {
+  // check common cases to avoid recursively constructing type in
+  // getMutableTypePtrImpl
+  auto kind = type->kind();
```
maybe this should be inlined into the mutableTypePtrHelper to avoid an extra indirection?

Commit:
Update on "[JIT] optimize mutableType calls"


I had previously tried to optimize getMutableTypePtr calls by not recursing through container types, but it turns out there are a few uses of container types which refine their contained elements.
This attempt was in #35301

Now I am optimizing calls by caching TypePtr -> Mutable TypePtr conversions. Now that we are doing caching none of the functions marked as const are really const anymore. Previously many of the const functions actually mutated internal state, such as rebuildWriteCache.

one kind of annoying thing is that there is a general api for querying mutability isMutableType that doesn't use the cache, and one internal that does, isMutableTypeInternal. It would be nice if I could call isMutableType within alias analysis and it would dispatch to the internal function, but I'm not sure how to do that.

getMutableTypePtr showed up as 12% of the first run of FairSeq, so this is a function worth optimizing.





[ghstack-poisoned]
